ipykernel/thread.py

```diff
@@ -14,29 +19,59 @@ class BaseThread(Thread):
     def __init__(self, **kwargs):
         """Initialize the thread."""
         super().__init__(**kwargs)
+        self.started = Event()
+        self.stopped = Event()
         self.pydev_do_not_trace = True
         self.is_pydev_daemon_thread = True
-        self.__stop = Event()
-        self._tasks_and_args: list[tuple[t.Any, t.Any]] = []
+        self._tasks: Queue[tuple[str, Callable[[], Awaitable[Any]]] | None] = Queue()
+        self._result: Queue[Any] = Queue()
+
+    @property
+    def task_group(self) -> TaskGroup:
+        return self._task_group
+
+    def start_soon(self, coro: Callable[[], Awaitable[Any]]) -> None:
+        self._tasks.put(("start_soon", coro))
 
-    def add_task(self, task: t.Any, *args: t.Any) -> None:
-        # May only add tasks before the thread is started.
-        self._tasks_and_args.append((task, args))
+    def run_async(self, coro: Callable[[], Awaitable[Any]]) -> Any:
+        self._tasks.put(("run_async", coro))
+        return self._result.get()
 
-    def run(self) -> t.Any:
+    def run_sync(self, func: Callable[..., Any]) -> Any:
+        self._tasks.put(("run_sync", func))
+        return self._result.get()
+
+    def run(self) -> None:
         """Run the thread."""
-        return run(self._main)
+        try:
+            run(self._main)
+        except Exception:
+            pass
```

Comment:
Do you want to at least log something if there is an exception ? It seems weird to me to swallow exception, but I'll trust you there.

Reply:
True, but I'm not sure if this would play well with threads like iostream, which does something with writing to `stdout`/`stderr`?
Maybe the exception could be made available through a method instead?